src/VNCVideoCapture.cpp

```diff
@@ -0,0 +1,206 @@
+#include <rfb/rfbclient.h>
+#include <signal.h>
+#include <time.h>
+#include "VNCVideoCapturer.h"
+
+#include "rtc_base/thread.h"
+#include "rtc_base/logging.h"
+#include "api/video_codecs/video_decoder.h"
+#include "media/base/videocapturer.h"
+#include "media/engine/internaldecoderfactory.h"
+#include "api/video/i420_buffer.h"
+#include "libyuv/video_common.h"
+#include "libyuv/convert.h"
+#include <iostream>
+
+char *host = "127.0.0.1";
+int port = 5900;
+
+static char * get_password (rfbClient *client) {
+  VNCVideoCapturer* that = (VNCVideoCapturer *) rfbClientGetClientData(client, (void *) "this");
+  return that->onGetPassword();
+}
+
+static void get_frame(rfbClient* client) {
+  VNCVideoCapturer* that = (VNCVideoCapturer *) rfbClientGetClientData(client, (void *) "this");
+  return that->onFrameBufferUpdate();
+}
+
+static void signal_handler(int sig) {
+  if (sig == SIGINT) {
+	RTC_LOG(LERROR) << __PRETTY_FUNCTION__ << "Recieved SIGINT, exiting ...";
+	exit(EXIT_FAILURE);
+  }
+}
+
+char* VNCVideoCapturer::onGetPassword() {
+  std::cout<< "Starting to get PASSWORD!!!" << std::endl;
+  // remove vnc:// prefix from uri before passing along
+  std::string url = uri.substr(6, std::string::npos);
+  // if it contains password, remove it from the URL
+  int index = url.find('@');
+  if (index != std::string::npos) {
+  	std::string password = url.substr(1, index - 1);
+	std::cout<< "Parsed Password: " << password << std::endl;
+    return strdup((char* )password.c_str());
+  }
+
+  if (getenv("VNCPASS") != NULL) {
+    return strdup(getenv("VNCPASS"));
+  }
+
+  this->onError("No password defined");
+  return NULL;
+}
+
+void VNCVideoCapturer::onClick(int x, int y, int buttonMask) {
+	double xRatio = x / 1000.0, yRatio = y / 1000.0;
+	RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "Sending click!!! (" << xRatio << ',' << yRatio << ") with click: " << buttonMask;
+	SendPointerEvent(client, (int) client->width * xRatio, (int) client->height * yRatio, buttonMask);
+}
+
+void VNCVideoCapturer::onPress(unsigned int code, bool down) {
+	RTC_LOG(LS_VERBOSE) << __PRETTY_FUNCTION__ << "Sending key!!! (" << code << ',' << down << ")";
+	SendKeyEvent(client, code, down);
+}
+
+void VNCVideoCapturer::onFrameBufferUpdate() {
```

Comment:
I'd be tempted to hoist the frame conversion code (~L89->L118?) into a separate function, could make it easier to test in isolation, but also separates out the logic of converting the frame format from encoding the video.